### DISCUSSION TITLE
Using message_queue=redis doesn't work

### DISCUSSION THREAD

**Discussion by liuqijie6** [miguelgrinberg/Flask-SocketIO#1937] [Q&A]
I'm currently trying to use gunicorn + flask-scoketIO for multiple jobs
but using message_queue=redis doesn't work

socketIo config
![image](https://user-images.githubusercontent.com/52588515/216504928-9db7610c-5275-42f3-ac13-973704b4dc67.png)

scoketIO server
![image](https://user-images.githubusercontent.com/52588515/216504600-667e26a1-449c-4552-995a-7d7313321a29.png)

test result: not print 123, not entering func

![image](https://user-images.githubusercontent.com/52588515/216504899-dee4591e-24e0-48d9-b332-d4d0bc303879.png)

**Comment by miguelgrinberg** [comment]
The `SocketIO` constructor is missing the `app` argument. How does this work at all?

**Reply by liuqijie6** [reply]
app.py
```
__all__ = ["app"]

import eventlet

eventlet.monkey_patch()
from __init__ import create_app
from units.extensions import socket_io
from config import CONFIG

app = create_app()

if __name__ == "__main__":
    print(app.url_map)
    host, port = CONFIG.get("web_server", "host"), CONFIG.get("web_server", "port")
    socket_io.run(app, host=host, port=port, debug=False, log_output=True)
```

**Reply by miguelgrinberg** [reply]
And now you are missing the message queue.

**Reply by liuqijie6** [reply]
sorry 。  i don't understand what you mean

**Reply by miguelgrinberg** [reply]
Can you show me your `extensions.py` file?

**Reply by liuqijie6** [reply]
extensions.py
```
from flask_executor import Executor
from flask_cors import CORS
from flask_socketio import SocketIO, join_room, leave_room, send
from flask_caching import Cache
from config import CONFIG

cache_config = {
    # "DEBUG": True,  # some Flask specific configs
    "CACHE_TYPE": "FileSystemCache",  # Flask-Caching related configs
    "CACHE_DEFAULT_TIMEOUT": 300,
    "CACHE_DIR": "cache",
    "CACHE_THRESHOLD": 2000,  # number of 'files' before start auto-delete

}

executor = Executor()
cors = CORS()
password, host, port = CONFIG.get("redis", "password"), CONFIG.get("redis", "host"), CONFIG.get("redis", "port")
socket_io = SocketIO(cors_allowed_origins='*',
                     async_mode="eventlet",  # eventlet  gevent # threading
                     engineio_logger=True,
                     logger=True,
                     # message_queue=f'redis://:{password}@{host}:{port}/0',
                      message_queue=f'redis://127.0.0.1:6379/0',
                     )
cache = Cache(config=cache_config)

```

**Reply by miguelgrinberg** [reply]
Okay. Try this:

```python
socket_io = SocketIO(app, cors_allowed_origins='*',
                     async_mode="eventlet",  # eventlet  gevent # threading
                     engineio_logger=True,
                     logger=True,
                     # message_queue=f'redis://:{password}@{host}:{port}/0',
                      message_queue=f'redis://127.0.0.1:6379/0',
                     )
```

I added `app` as an argument, I mentioned this in my first response and you apparently ignored it.

**Reply by liuqijie6** [reply]
**Is it the same as what you said when I write this way?**
**but not working**
app.py
![image](https://user-images.githubusercontent.com/52588515/216860033-395d2e55-2820-4d82-90bc-6510c00d0224.png)

**Reply by miguelgrinberg** [reply]
No, it is not the same. Please make the change I suggested exactly as I put it in my example.

**Reply by liuqijie6** [reply]
I use scoketIO's factory pattern. If you write as you said, it will lead to a circular import error.

**Reply by liuqijie6** [reply]
@miguelgrinberg  If I comment out message_queue it works fine.

**Reply by miguelgrinberg** [reply]
The `app` argument is required. If you get a circular import you have to fix that. You can't omit this argument. Tried to say this several times already and you keep ignoring it.

**Comment by liuqijie6** [comment]
I try. but dont work.

__init__.py
```
__all__ = ["create_app", "BASE_DIR", "flask_func"]

import os
import sys


def flask_func():
    from flask import Flask
    return Flask(__name__)


def create_app():
    from routers import index
    from units.extensions import (executor, cors, cache)
    from routers.soketIo.socketIo_server import socket_io

    app = flask_func()

    """ flask 配置
    https://dormousehole.readthedocs.io/en/latest/config.html
    """
    app.config['EXECUTOR_PROPAGATE_EXCEPTIONS'] = True
    app.config['EXECUTOR_MAX_WORKERS'] = 5
    app.config['EXECUTOR_TYPE'] = 'thread'
    app.config['JSON_AS_ASCII'] = True
    # app.config['JSONIFY_PRETTYPRINT_REGULAR'] = True
    # app.config['UPLOAD_FOLDER'] = f'{os.path.dirname(__file__)}/static/'
    """flask 蓝图注册"""
    app.register_blueprint(index)

    """flask 扩展"""
    executor.init_app(app)
    cors.init_app(app)
    socket_io.init_app(app)
    cache.init_app(app)

    return app

```

app.py
```
__all__ = ["app"]

import eventlet

eventlet.monkey_patch()
from __init__ import create_app
from units.extensions import socket_io
from config import CONFIG

app = create_app()

if __name__ == "__main__":
    print(app.url_map)
    host, port = CONFIG.get("web_server", "host"), CONFIG.get("web_server", "port")
    socket_io.run(app, host=host, port=port, debug=False, log_output=True)

```
units/extensions.py
```
# __all__ = ["executor", "cors", "socket_io", "cache"]
""" flask 扩展"""

from flask_executor import Executor
from flask_cors import CORS
from flask_socketio import SocketIO, join_room, leave_room, send
from flask_caching import Cache
from config import CONFIG

cache_config = {
    # "DEBUG": True,  # some Flask specific configs
    "CACHE_TYPE": "FileSystemCache",  # Flask-Caching related configs
    "CACHE_DEFAULT_TIMEOUT": 300,
    "CACHE_DIR": "cache",
    "CACHE_THRESHOLD": 2000,  # number of 'files' before start auto-delete

}

executor = Executor()
cors = CORS()
password, host, port = CONFIG.get("redis", "password"), CONFIG.get("redis", "host"), CONFIG.get("redis", "port")

from __init__ import flask_func

socket_io = SocketIO(app=flask_func(), cors_allowed_origins='*',
                     async_mode="eventlet",  # eventlet  gevent # threading
                     engineio_logger=True,
                     logger=True,
                     # message_queue=f'redis://:{password}@{host}:{port}/0',
                     message_queue=f'redis://127.0.0.1:6379/0',
                     )
cache = Cache(config=cache_config)

```

output, but no print.
![image](https://user-images.githubusercontent.com/52588515/217127094-e314d2a1-6cd5-464e-9cc7-fe597ab33b8c.png)

**Reply by miguelgrinberg** [reply]
I think you need to spend some time working with simpler application structures. When I said that you have to pass the app instance I did not mean that you have to create a different app instance and use that. I meant that you need to pass the main application instance, the one that you use for your Flask application routes.

I suggest you work with some simpler examples to understand how to structure your Flask application, because what you are doing is extremely messy and prone to errors.

Start from https://github.com/miguelgrinberg/Flask-SocketIO-Chat for example. Then if you want to see something a bit more advanced, try https://github.com/miguelgrinberg/flack.

**Reply by liuqijie6** [reply]
thanks

**Reply by liuqijie6** [reply]
# https://github.com/miguelgrinberg/Flask-SocketIO-Chat
I have a question . Why don't you write `SocketIO(app)` here as you said above?
<img width="491" alt="image" src="https://user-images.githubusercontent.com/52588515/217279676-0f884291-2432-4a4d-b389-57e2ae5cc06f.png">

<img width="465" alt="image" src="https://user-images.githubusercontent.com/52588515/217279819-62ec1cbc-dc99-49be-923a-d14cc9f8137c.png">


Currently my writing is very similar to this。

**Reply by miguelgrinberg** [reply]
This is standard for all Flask extensions. You initialize them with no arguments, then pass the arguments in the `init_app()` call.
